src/Playwright/Core/HarRouter.cs

```diff
@@ -0,0 +1,152 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+using System;
+using System.Linq;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using Microsoft.Playwright.Helpers;
+
+namespace Microsoft.Playwright.Core
+{
+    internal class HarRouter
+    {
+        private readonly LocalUtils _localUtils;
+
+        private readonly string _harId;
+
+        private readonly HarRouterOptions _options;
+
+        private readonly HarNotFound _notFoundAction;
+
+        private HarRouter(LocalUtils localUtils, string harId, HarNotFound notFoundAction, HarRouterOptions options)
+        {
+            _localUtils = localUtils;
+            _harId = harId;
+            _options = options;
+            _notFoundAction = notFoundAction;
+        }
+
+        internal static async Task<HarRouter> CreateAsync(LocalUtils localUtils, string file, HarNotFound? notFoundAction, HarRouterOptions options)
+        {
+            var (harId, error) = await localUtils.HarOpenAsync(file).ConfigureAwait(false);
+            if (!string.IsNullOrEmpty(error))
+            {
+                throw new PlaywrightException(error);
+            }
+            return new HarRouter(localUtils, harId, notFoundAction ?? HarNotFound.Abort, options);
+        }
+
+        private async Task HandleAsync(Route route)
+        {
+            var request = route.Request;
+            var response = await _localUtils.HarLookupAsync(
+                harId: _harId,
+                url: request.Url,
+                method: request.Method,
+                headers: (await request.HeadersArrayAsync().ConfigureAwait(false)).ToList(),
+                postData: request.PostDataBuffer,
+                isNavigationRequest: request.IsNavigationRequest).ConfigureAwait(false);
+
+            if (response.Action == "redirect")
+            {
+                await route.RedirectNavigationRequestAsync(response.RedirectURL).ConfigureAwait(false);
+                return;
+            }
+            if (response.Action == "fulfill")
+            {
+                await route.FulfillAsync(new()
+                {
+                    Status = response.Status,
+                    Headers = new RawHeaders(response.Headers).Headers,
+                    BodyBytes = response.Body,
+                }).ConfigureAwait(false);
+                return;
+            }
+
+            if (_notFoundAction == HarNotFound.Abort)
+            {
+                await route.AbortAsync().ConfigureAwait(false);
+                return;
+            }
+
```

Comment:
Put the no-op error placeholder just to keep things parallel to the other ports.